torch_xla/core/xla_model.py

```diff
@@ -615,19 +620,100 @@ def all_gather(value, dim=0, groups=None, output=None, pin_layout=True):
-          output, value, token, dim, shard_count, groups or [], pin_layout)
-      torch_xla._XLAC._set_all_reduce_token(devctx.device, new_token)
-      return output
-
-    result = torch_xla._XLAC._xla_all_gather_coalesced(value, token, dim,
-                                                       shard_count, groups or
-                                                       [], pin_layout)
-    torch_xla._XLAC._set_all_reduce_token(devctx.device, result[-1])
-    return result[:-1]
+        raise ValueError("`output` length doesn't match `value` length: "
+                         f"{len(output)} vs {len(value)}.")
+
+    # helper function for bucketing
+    def _all_gather_coalesced(tensor_list, output_list=None):
+      if output_list != None:
+        if not isinstance(output_list, list) or any(
+            not isinstance(v, torch.Tensor) for v in output_list):
+          raise TypeError(
+              f"_all_gather_coalesced: `output_list` needs to be a list of Tensors, but given {type(output_list)}."
+          )
+        if len(output_list) != len(tensor_list):
+          raise ValueError(
+              "_all_gather_coalesced: `output_list` length doesn't match `tensor_list` length: "
+              f"{len(output_list)} vs {len(tensor_list)}.")
+        # Call the out of place version of the reduce_scatter
+        new_token = torch_xla._XLAC._xla_all_gather_coalesced_out(
+            output_list, tensor_list, token, dim, shard_count, groups or [],
+            pin_layout)
+        torch_xla._XLAC._set_all_reduce_token(devctx.device, new_token)
+        return output_list
+
+      result = torch_xla._XLAC._xla_all_gather_coalesced(
+          tensor_list, token, dim, shard_count, groups or [], pin_layout)
+      torch_xla._XLAC._set_all_reduce_token(devctx.device, result[-1])
+      return result[:-1]
+
+    total = 0
+    tensor_bucket = []
+    output_bucket = []
+    out_tensors = []
+    bucket_cap = int(
+        os.getenv("ALL_GATHER_REDUCE_SCATTER_BUCKET_CAP_MB",
+                  _ALL_GATHER_REDUCE_SCATTER_BUCKET_CAP_MB)) * 1024 * 1024
+    if groups:
+      divisor = len(groups[0]) if type(groups[0]) == list else len(groups)
+    else:
+      divisor = xrt_world_size()
+    bucket_cap = bucket_cap / divisor
+
+    for idx, tensor in enumerate(value):
+      tensor_bytes = tensor.numel() * tensor.element_size()
+      output_selected = None
+      if output != None:
+        output_selected = output[idx]
+        if tensor.numel() != output_selected.numel():
+          raise ValueError(
+              f"`output` tensor size doesn't match `input` tensor size for tensor list index {idx}: "
+              f"{output_selected.numel()} vs {tensor.numel()}.")
+
+      # Tensor is larger than bucket_cap, don't bucketize
+      if tensor_bytes > bucket_cap:
+        # Flush out previous buckets even if they don't fill up
+        if total >= 0.5 * bucket_cap or (total + tensor_bytes) > 2 * bucket_cap:
```

Comment:
@aws-rhsoln in this if statement, if tensor_bytes is 500MB and bucket_cap is 160MB, event of total is 0 because buckets are empty, it would try to call _all_gather_coalesced on empty buckets, and crash?

Reply:
Also there are 2 magic numbers 0.5 and 2, making it difficult to test/reason. Again, we should remove.